### PR TITLE
[Gradient Compression] Refactor default_hooks.py and powerSGD_hook.py by creating a util function that make a vanilla allreduce future

### PR DESCRIPTION
**Pull request by wayi1**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#51094 [Gradient Compression] Refactor default_hooks.py and powerSGD_hook.py by creating a util function that make a vanilla allreduce future**
* #50985 [Gradient Compression] Explicitly specify the dtype of the error tensor
* #50981 [Gradient Compression] Simplify the implementation of warm-start
* #50974 [Gradient Compression] Typo fixes in PowerSGD
* #50973 [Gradient Compression] Allow PowerSGD to run vallina allreduce for the first K iterations

Address https://github.com/pytorch/pytorch/pull/50973#discussion_r564229818

Original PR issue: Investigate Applying PowerSGD to Communication Hook for Gradient Compression #47202

Differential Revision: [D26070147](https://our.internmc.facebook.com/intern/diff/D26070147/)

### CODE REVIEW

File: torch/distributed/algorithms/ddp_comm_hooks/default_hooks.py [hunk 1]
```diff
@@ -2,6 +2,18 @@
 import torch.distributed as dist
 
 
+def allreduce_fut(
```
Comment (rohan-varma) [inline torch/distributed/algorithms/ddp_comm_hooks/default_hooks.py:5]: Can we add a test ensuring that this allreduce is equivalent to DDP allreduce? Or do you think this is already well covered by current tests?

File: torch/distributed/algorithms/ddp_comm_hooks/default_hooks.py [hunk 2]
```diff
@@ -2,6 +2,18 @@
...
+    process_group: dist.ProcessGroup, tensor: torch.Tensor
+) -> torch.futures.Future:
+    "Averages the input gradient tensor by allreduce and returns a future."
+    fut = dist.all_reduce(tensor, group=process_group, async_op=True).get_future()
+
+    def div_by_group_size(fut):
+        return [fut.value()[0].div_(process_group.size())]
```
Comment (rohan-varma) [inline torch/distributed/algorithms/ddp_comm_hooks/default_hooks.py:12]: Is it  possible for `process_group` to be `None`? Should we add an assert if not?

**Review by rohan-varma — APPROVED** [review]

File: torch/distributed/algorithms/ddp_comm_hooks/default_hooks.py (hunk 2 above)
Comment (SciPioneer) [reply torch/distributed/algorithms/ddp_comm_hooks/default_hooks.py:12]: Good idea! Duplicated `group_to_use = process_group if process_group is not None else dist.group.WORLD` inside this util function.

File: torch/distributed/algorithms/ddp_comm_hooks/default_hooks.py (hunk 1 above)
Comment (SciPioneer) [reply torch/distributed/algorithms/ddp_comm_hooks/default_hooks.py:5]: This is more like an internal util function, and both `allreduce_hook` and `powerSGD_hook` are the public APIs that call this util function are tested. I don't think we need a separate test for this util function, as our tests don't need to cover each private/internal function.

**Comment by izdeby** [comment]
Reverting due to a broken CI

**Comment by rohan-varma** [comment]
@SciPioneer Looks like the failures on this PR were legit:

```
Jan 29 01:19:21 Traceback (most recent call last):
Jan 29 01:19:21   File "distributed/test_c10d.py", line 21, in <module>
Jan 29 01:19:21     import torch.distributed.algorithms.ddp_comm_hooks.default_hooks as default
Jan 29 01:19:21   File "/opt/conda/lib/python3.6/site-packages/torch/distributed/algorithms/ddp_comm_hooks/__init__.py", line 7, in <module>
Jan 29 01:19:21     from . import (
Jan 29 01:19:21   File "/opt/conda/lib/python3.6/site-packages/torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py", line 7, in <module>
Jan 29 01:19:21     import torch.distributed.algorithms.ddp_comm_hooks.default_hooks as default
Jan 29 01:19:21 AttributeError: module 'torch.distributed.algorithms' has no attribute 'ddp_comm_hooks'
Jan 29 01:19:21 Traceback (most recent call last):
Jan 29 01:19:21   File "test/run_test.py", line 922, in <module>
Jan 29 01:19:21     main()
Jan 29 01:19:21   File "test/run_test.py", line 901, in main
Jan 29 01:19:21     raise RuntimeError(err_message)
Jan 29 01:19:21 RuntimeError: distributed/test_c10d failed!
Jan 29 01:19:22 + cleanup
Jan 29 01:19:22 + retcode=1
Jan 29 01:19:22 + set +x
```

**Pull request closed without merging**